src/flp/types/fixedpoint_l2.rs

```diff
@@ -0,0 +1,816 @@
+
+        // Ensure that all submitted field elements are either 0 or 1.
+        // This is done for:
+        //  - all vector entries (each of them encoded in `self.bits_per_entry`
+        //    field elements)
+        //  - the submitted norm (encoded in `self.bits_for_norm` field
+        //    elements)
+        //
+        // Since all input vector entry (field-)bits, as well as the norm bits,
+        // are contiguous, we do the check directly for all bits from 0 to
+        // entries*bits_per_entry + bits_for_norm.
+        //
+        // In order to keep the proof size down, this is done using the
+        // `ParallelSum` gadget. For a similar application see the `CountVec`
+        // prio3 type.
+        //
+        // Check that each element is a 0 or 1:
+        let mut r = joint_rand[0];
+        let mut range_check = F::zero();
+        let mut padded_chunk = vec![F::zero(); 2 * self.gadget0_chunk_len];
+        for chunk in input[..self.range_norm_end].chunks(self.gadget0_chunk_len) {
+            let d = chunk.len();
+            for i in 0..self.gadget0_chunk_len {
+                if i < d {
+                    padded_chunk[2 * i] = chunk[i];
+                } else {
+                    // If the chunk is smaller than the chunk length, then copy the last element of
+                    // the chunk into the remaining slots.
+                    padded_chunk[2 * i] = chunk[d - 1];
+                }
+                padded_chunk[2 * i + 1] = r * constant_part_multiplier;
+                r *= joint_rand[0];
+            }
+
+            range_check += g[0].call(&padded_chunk)?;
+        }
+
+        // Compute the norm of the entries and ensure that it is the same as the
+        // submitted norm. There are exactly enough bits such that a submitted
+        // norm is always a valid norm (semantically in the range [0,1]). By
+        // comparing submitted with actual, we make sure the actual norm is
+        // valid.
+        //
+        // The function to compute here (see explanatory comment at the top) is
+        //   norm(ys) = sum_{y in ys} y^2 - (2^n)*y + 2^(2n-2)
+        //
+        // This is done by the `ParallelSum` gadget `g[1]`, which evaluates the
+        // inner polynomial on each (decoded) vector entry, and then sums the
+        // results. Note that the gadget is not called on the whole vector at
+        // once, but sequentially on chunks of size `self.gadget1_chunk_len` of
+        // it. The results of these calls are accumulated in the `outp` variable.
+        //
+        // decode the bit-encoded entries into elements in the range [0,2^n):
+        let decoded_entries: Result<Vec<_>, _> = input[0..self.entries * self.bits_per_entry]
+            .chunks(self.bits_per_entry)
+            .map(F::decode_from_bitvector_representation)
+            .collect();
+
+        // run parallel sum gadget on the decoded entries
+        let computed_norm = {
```

Comment:
Readability: I would expect this code block to look more or less like the computation of `range_check` above. This could we be a lot more readable if they had the same overall structure.

Reply:
it's a bit different than the one above, as the norm computation does not use BlindPolyEval so the chunk structure is simpler. i added a comment to explain and adjusted things a bit.